add support for Java 8 completions (promises) API in FFmpeg.executeAsync()

guss77 opened the pull request:
With the unfortunate state of Java `Future` and `FutureTask` - an API introduced in Java 5, before explosion in popularity of the "promise"-style APIs - that is used in `FFmpeg.executeAsync()`, it is not possible to wait for an asynchronous task completion without blocking a thread.

With Java 8's introduction of the `CompletionStage` interface and `CompletableFuture` implementation, this has changed and allows a simpler way to wait for completion without blocking, that maps well to established asynchronous and reactive programming patterns.

This PR adds into `FFmpegResultFuture` support for producing a `CompletableFuture` instance so that additional actions can be chained to the processing result without blocking an additional thread. Chained actions are called in the context of the `executeAsync()`'s internal runner thread, right before it terminates, so the additional behavior should not cause any scheduling conflict or increased latency (other than expected).

kokorin commented:
Hello @guss77,
first of all thank for the PR and sorry for the delay.

I think another approach is also possible: 

1. change `FFmpegResultFuture#resultFuture` property's type from `Future<FFmpegResult>` to `CompletableFuture<FFmpegResult>`
2. make `FFmpegResultFuture` to implement `CompletionStage`  and delegate all methods to `resultFuture` property.

Or to make it even more simple we can just go with option 1 and  `FFmpegResultFuture#toCompletableFuture` method.

What do you think?

guss77 commented:
Thanks @kokorin and no worries (actually, 2 days is a very decent response time in F/LOSS 😅😎).

This initial PR was mostly about being very minimal and prompting discussion, so thanks for that!

1. I like this approach, but implementing it will require 2 major changes:
  * add a public getter for `FFmpegResultFuture#resultFuture` which is currently missing (possibly dropping all the `resultFuture` delegates).
  * Change `FFmpeg#executeAsync()` to execute `ProcessHandler` in another way as it currently relies on the `FutureTask` implementation of `Future` which conflicts with `FutureTask`. Which may also be a good time to have a discussion about thread management and should the `ProcessHandler` be executed on the common fork/join pool or another `Executor` (possibly controlled by the user).
2. Possibly the Right Thing™️ and requires a lot less policy changes, but it does need a good chunk of code for the delegation. I actually started going that route, saw how many LoCs are needed and decided not to start contributing to a new project by submitting a >300 LoC patch 😝.
3. Probably the best option I think, though it is basically the same as option 1 - it just uses a specific name for the getter.

I think the "Correct" approach would be to do all of these:

1. Use `CompletableFuture` as the internal implementation
2. Expose it using `toCompletableFuture()`, which is identical to the the `CompletionStage.toCompletableFuture()` API.
3. While were at it, implement the `CompletionStage` interface as it is just boilerplate to get it to play nice with friends (BTW - the `CompletableFuture.toCompletableFuture()` implementation is just `return this` so it is identical to the getter from step 2).

If that is OK with you, I can push such an implementation here.

kokorin commented:
@guss77 I mostly agree with the "Correct" approach, but still not sure if it's required to implement `CompletionStage`. 

On the other side it makes sense to extend `CompletableFuture` and delegate `cancel()` method invocations to the `Stopper`. I didn't extend `Future` because one of its cancel methods `cancel(boolean)`  are not useful if invoked with `false` value. And hence parameter-less cancel method has no sense too.

`CompletableFuture` JavaDoc says: "Method cancel has the same effect as completeExceptionally(new CancellationException())" Probably it makes sense to keep this behavior.  

One more problem that has came to my mind is: will ffmpeg be stopped indeed when `CompletableFuture#cancel()` is invoked? Or will it continue its work?

While thinking about it I drafted the following snippet:

```java
return new FFmpegResultFuture(
        CompletableFuture.supplyAsync(
                processHandler::execute,
                Executors.newSingleThreadExecutor(runnable -> {
                    Thread runner = new Thread(runnable, "FFmpeg-async-runner");
                    runner.setDaemon(true);
                    return runner;
                })
        ),
        stopper
);
```

Probably it's a mess and should be refactored.

Regarding PR: please continue. It would great if you:
1. add yourself to the authors list in Apache license in files you touch
2. add working example near to other examples
3. add shortened example to Readme

guss77 commented:
I'll offer something tomorrow (today is done for me).

A note about `Executros`: You can't fire one-off executors - you have to stop them when you're done with them (or not, it is often fine if you keep one until the end of the VM), otherwise they accumulate. They don't garbage collect.

kokorin commented:
@guss77 Executors are collected, check `FinalizableDelegatedExecutorService`. In short it uses `finalize()` method to shutdown wrapped executor.

guss77 commented:
@kokorin, yes - but `finalize()` is not guaranteed to ever run and the JVM notoriously doesn't reap non-young objects until absolutely required (and unless you are running ffmpeg on really short files, the supporting objects will be promoted) and unlike a few bits of memory, the resource consumption of a bunch unused tenured executors with threads could be serious. Spawning executors is generally frowned upon.

I think if we're going into executors, it is a great API to let the user control the scheduling of tasks - so it would be a good API to expose to user control. I'll put something in the patch I'm working on now.

guss77 commented:
Also, regarding `Future.cancel(boolean)` - isn't this exactly what `FFmpegResultFuture.stop(boolean)` does? Why not implement it identically?

kokorin commented:
Of course I agree with you regarding `finilize()` and guarantees that JVM gives about it. Probably we can solve this issue by custom `Executor` implementation which doesn't store `Thread` in a pool, but starts new `Thread` every time.

> 
> Also, regarding `Future.cancel(boolean)` - isn't this exactly what `FFmpegResultFuture.stop(boolean)` does? Why not implement it identically?

It exactly does what you describe. But if `FFmpegResultFuture` extends `Future` then Liskov principle will be broken: `FFmpegResultFuture` can be uses as `Future` but `Future#cancel(mayInterrupt=false)` will stop running task, while contract demands that it can cancel only a task which hasn't been started yet.

guss77 commented:
> Future#cancel(mayInterrupt=false)` will stop running task, while contract demands that it can cancel only a task which hasn't been started yet.

I'm not sure why you say that - my reading of [the contract of `Future#cancel(boolean)`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Future.html#cancel-boolean-) is that sending `false` while the task is running will complete the task immediately to a "cancelled" state without (necessarily) stopping the task - so it should definitely cancel a task that has started. That is just the straightforward reading of the Javadoc.

IMHO, the value of `mayInterrupt` is more of a hint than an actual contract as whether or not anything has been interrupted or not is an implementation-specific behavior as no side effects can be observed anyway (if cancellation succeeded then `get()` must throw a `CancellationException`).

guss77 commented:
Notes about the last push:
- I implemented stopping using the `CompletableFuture#cancel(boolean)` API, so I also removed the stopper from `FFmpegResultFuture` and implemented stopping by calling cancel.
- I added an `executeAsync(Executor)` API so the user can control (some of) the thread scheduling, and moved the "one async thread" into an `Executor` interface implementation - so it should work identically to how it was before without getting into the which executor question (this is basically how `CompletableFuture` does it anyway).

There's probably a few more points that can use cleaning up or simplifying around this patch, but please let me know what you think about this one, first.

guss77 commented:
Unfortunately, the last change broke the stop unit test. I'm working on it.

guss77 commented:
> Unfortunately, the last change broke the stop unit test. I'm working on it.

Ok, the problem is related to the behavior of the stop/cancel functionality (at `FFmpegTest.testGraceAsyncStop()`) - according to the [`Future` API](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Future.html), cancelling would cause `get()` to throw a `CancellationException`, and this is indeed what `FFmpegResult.get(...)` documents (even though it does not implement `Future`).

But `FFmpegTest.testGraceAsyncStop()` assume that it doesn't and will fail if after calling `graceStop()` the `get(...)` call throws.

I think maintaining both the test as is (i.e. the current behavior of `get(...)` after stopping) as well as the documentation is untenable, but maintaining the current behavior on top of the `CompletableFuture` API is going to take some fancy leg work (read "hacky leg work" - I can backup the result in another reference to be used by `get()` if it catches a cancel exception).

kokorin commented:
Hello, sorry for long reply.

> IMHO, the value of `mayInterrupt` is more of a hint than an actual contract as whether or not anything has been interrupted or not is an implementation-specific behavior as no side effects can be observed anyway (if cancellation succeeded then `get()` must throw a `CancellationException`).

Grace stop was implemented almost year ago, please check #81 and #85 

Sometimes its required to stop ffmpeg gracefully. For example during live stream recording or desktop recording. Record duration is not know before. When user decides to stop recording output file should not be corrupted and hence ffmpeg execution should be considered successful. 

So `FFmpegResultFuture` should provide a method for grace stop other then `cancel(false)`. It should use `Stopper` to stop ffmpeg, not thread interruption.

guss77 commented:
> Hello, sorry for long reply.

NP

> Sometimes its required to stop ffmpeg gracefully. For example during live stream recording or desktop recording. Record duration is not know before. When user decides to stop recording output file should not be corrupted and hence ffmpeg execution should be considered successful.

Ah, right. I didn't consider this use case.

> So `FFmpegResultFuture` should provide a method for grace stop other then `cancel(false)`. It should use `Stopper` to stop ffmpeg, not thread interruption.

The current PR uses the stopper without thread interruption (a deprecated feature anyway), but I'll put back graceful interruption.

guss77 commented:
The latest version of the PR implements `FFmpegResultFuture.graceStop()` by calling the stopper - and this would cause the completion to complete as expected with the partial result, while `FFmpegResultFuture.forceStop()` is implemented by cancelling the completion - that will in turn `forceStop` the stopper and cause the promise to throw a `CancellationException` when trying to retrieve its value.

kokorin commented on src/main/java/com/github/kokorin/jaffree/ffmpeg/FFmpeg.java:
```diff
@@ -363,20 +365,49 @@ public FFmpegResult execute() {
+     * @param executor the executor to use for asynchronous execution
+     * @return ffmpeg result future
+     */
+    public FFmpegResultFuture executeAsync(java.util.concurrent.Executor executor) {
```
I would suggest to make this method protected. Please, import `java.util.concurrent.Executor`.

kokorin commented on src/main/java/com/github/kokorin/jaffree/ffmpeg/FFmpeg.java:
```diff
@@ -363,20 +365,49 @@ public FFmpegResult execute() {
-        Thread runner = new Thread(resultFuture, "FFmpeg-async-runner");
-        runner.setDaemon(true);
-        runner.start();
+        Objects.requireNonNull(executor, "Executor service must be provided").execute(new Runnable() {
```
Parameters should be verified in the beginning. Personally I would omit this check.
The string is too long, please run checkstyle.

kokorin commented:
Please pay attention to the following:
1. PR should be created to develop branch, not master (preferably from feature branch)
2. add yourself to the authors list in Apache license in files you touch
3. add working example near to other examples
4. add shortened example to Readme

guss77 commented:
>     1. PR should be created to develop branch, not master (preferably from feature branch)

I changed the base, but apparently I can't change the source. I'll know to use a feature branch next time.

kokorin commented:
@guss77 please, rebase your changes on develop branch. This PR now contains merge commits from master branch.
Probably new PR has to be created.

guss77 commented:
Yea, it's a mess now. I'll close this PR and redo it properly (a bit later - I'm going on a short vacation now).

Thank you for your patience! 👍😅
